what4/src/What4/Utils/BVDomain.hs

```diff
@@ -1,18 +1,503 @@
+  where
+    -- Invariant: the gap between @b@ and @c@ is the biggest we've
+    -- seen between adjacent values so far.
+    go (a, b) (c, d) [] = union (range w a b) (range w c d)
+    go (a, b) (c, d) (e : rest)
+      | e - d > c - b = go (a, d) (e, e) rest
+      | otherwise     = go (a, b) (c, e) rest
+
+-- | Return union of two domains.
+union :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
+union a b =
+  case a of
+    BVDAny _ -> a
+    BVDInterval _ al aw ->
+      case b of
+        BVDAny _ -> b
+        BVDInterval mask bl bw ->
+          interval mask cl (ch - cl)
+          where
+            size = mask + 1
+            ac = 2 * al + aw
+            bc = 2 * bl + bw
+            al' = if ac + mask < bc then al + size else al
+            bl' = if bc + mask < ac then bl + size else bl
+            ah' = al' + aw
+            bh' = bl' + bw
+            cl = min al' bl'
+            ch = max ah' bh'
+
+-- | @concat a y@ returns domain where each element in @a@ has been
+-- concatenated with an element in @y@.  The most-significant bits
+-- are @a@, and the least significant bits are @y@.
+concat :: NatRepr u -> BVDomain u -> NatRepr v -> BVDomain v -> BVDomain (u + v)
+concat u a v b =
+  case a of
+    BVDAny _ -> BVDAny mask
+    BVDInterval _ al aw -> interval mask (cat al bl) (cat aw bw)
+  where
+    cat i j = i `shiftL` widthVal v + j
+    mask = maxUnsigned (addNat u v)
+    (bl, bh) = ubounds b
+    bw = bh - bl
+
+-- | @shrink i a@ drops the @i@ least significant bits from @a@.
+shrink ::
+  NatRepr i ->
+  BVDomain (i + n) -> BVDomain n
+shrink i a =
+  case a of
+    BVDAny mask -> BVDAny (shr mask)
+    BVDInterval mask al aw ->
+      interval (shr mask) bl (bh - bl)
+      where
+        bl = shr al
+        bh = shr (al + aw)
+  where
+    shr x = x `shiftR` widthVal i
+
+-- | @trunc n d@ selects the @n@ least significant bits from @d@.
+trunc ::
+  (n <= w) =>
+  NatRepr n ->
+  BVDomain w -> BVDomain n
+trunc n a =
+  case a of
+    BVDAny _ -> BVDAny mask
+    BVDInterval _ al aw -> interval mask al aw
+  where
+    mask = maxUnsigned n
+
+-- | @select i n a@ selects @n@ bits starting from index @i@ from @a@.
+select ::
+  (1 <= n, i + n <= w) =>
+  NatRepr i ->
+  NatRepr n ->
+  BVDomain w -> BVDomain n
+select i n a = shrink i (trunc (addNat i n) a)
+
+zext :: (1 <= w, w+1 <= u) => BVDomain w -> NatRepr u -> BVDomain u
+zext a u = range u al ah
+  where (al, ah) = ubounds a
+
+sext ::
+  forall w u. (1 <= w, w + 1 <= u) =>
+  NatRepr w ->
+  BVDomain w ->
+  NatRepr u ->
+  BVDomain u
+sext w a u =
+  case fProof of
+    LeqProof ->
+      range u al ah
+      where (al, ah) = sbounds w a
+  where
+    wProof :: LeqProof 1 w
+    wProof = LeqProof
+    uProof :: LeqProof (w+1) u
+    uProof = LeqProof
+    fProof :: LeqProof 1 u
+    fProof = leqTrans (leqAdd wProof (knownNat :: NatRepr 1)) uProof
+
+--------------------------------------------------------------------------------
+-- Shifts
+
```

Comment:
I feel like we can probably do something reasonable with the shift operations

Reply:
Yes, I was just lazy and my goal was feature parity with the old one (which didn't implement these either). I'll go ahead and implement those.

Reply:
`lshr` and `ashr` are done in 9cdc00333ada1ae8ce7ef0e459706d8b004221be.

Reply:
... and now `shl` is done in 013337e829aba60a996d5a27c2b7ee03e2eb996f. (This one was a lot trickier!)